pkg/internal/traces/read_decorator.go

```diff
@@ -0,0 +1,99 @@
+type InstanceIDConfig struct {
+	// HostnameDNSResolution is true if Beyla uses the DNS to resolve the local hostname or
+	// false if it uses the local hostname.
+	HostnameDNSResolution bool `yaml:"dns" env:"BEYLA_HOSTNAME_DNS_RESOLUTION"`
+	// OverrideHostname can be optionally set to avoid resolving any hostname and using this
+	// value. Beyla will anyway attach the process ID to the given hostname for composing
+	// the instance ID.
+	OverrideHostname string `yaml:"override_hostname" env:"BEYLA_HOSTNAME"`
+	// OverrideInstanceID can be optionally set to avoid Beyla composing the instance ID
+	// and use this value. If you are managing multiple processes from a single Beyla instance,
+	// all the processes will have the same Instance ID.
+	OverrideInstanceID string `yaml:"override_instance_id" env:"BEYLA_INSTANCE_ID"`
+}
+
+// ReadDecorator is the input node of the processing graph. The eBPF tracers will send their
+// traces to the ReadDecorator's TracesInput, and the ReadDecorator will decorate the traces with some
+// basic information (e.g. instance ID) and forward them to the next pipeline stage
+type ReadDecorator struct {
+	TracesInput <-chan []request.Span
+
+	InstanceID InstanceIDConfig
+}
+
+// decorator modifies a []request.Span slice to fill it with extra information that is not provided
+// by the tracers (for example, the instance ID)
+type decorator func(spans []request.Span)
+
+func ReadFromChannel(ctx context.Context, r ReadDecorator) (node.StartFunc[[]request.Span], error) {
+	decorate := getDecorator(&r.InstanceID)
+	return func(out chan<- []request.Span) {
+		cancelChan := ctx.Done()
+		for {
+			select {
+			case trace, ok := <-r.TracesInput:
+				if ok {
+					decorate(trace)
+					out <- trace
+				} else {
+					rlog().Debug("input channel closed. Exiting traces input loop")
+					return
+				}
+			case <-cancelChan:
+				rlog().Debug("context canceled. Exiting traces input loop")
+				return
+			}
+		}
+	}, nil
+}
+
+func getDecorator(cfg *InstanceIDConfig) decorator {
+	if cfg.OverrideInstanceID != "" {
+		return func(spans []request.Span) {
+			for i := range spans {
+				spans[i].ServiceID.Instance = cfg.OverrideInstanceID
+			}
+		}
+	}
+	return hostNamePIDDecorator(cfg)
+}
+
+func hostNamePIDDecorator(cfg *InstanceIDConfig) decorator {
+	// TODO: periodically update in case the current Beyla instance is created from a VM snapshot running as a different hostname
+	resolver := hostname.CreateResolver(cfg.OverrideHostname, "", cfg.HostnameDNSResolution)
+	fullHostName, _, err := resolver.Query()
+	log := rlog().With("function", "instance_ID_hostNamePIDDecorator")
+	if err != nil {
+		log.Warn("can't read hostname. Leaving empty. Consider overriding"+
+			" the BEYLA_HOSTNAME or BEYLA_INSTANCE_ID properties",
+			"error", err)
+	} else {
+		log.Info("using hostname", "hostname", fullHostName)
+	}
+
+	return func(spans []request.Span) {
+		for i := range spans {
```

Comment:
I think the sprintf won't be too slow, if that's the only thing we run through here, but it might be nice to have an LRU cache, since we expect events to mostly come from few PIDs?

Reply:
Yeah! I actually left the TO-DO below. I'll merge this PR and add the cache into another PR.

I'll first check what is faster, managing the LRU cachd & hashmap or maybe just replace this Sprintf by a  hostname + "-" + Atoi(pid).